### PR TITLE
Do not panic when a default snapshotclass does not exist

### LINKED ISSUE #1203: Using volumegroup snapshot requires a volumesnapshotclass?
I was attempting to use groupsnapshots but I forgot to create `volumesnapshotclass`. 

1. Here is my groupsnapshotclass:

```
---
apiVersion: groupsnapshot.storage.k8s.io/v1beta1
kind: VolumeGroupSnapshotClass
metadata:
  name: csi-hostpath-groupsnapclass
driver: hostpath.csi.k8s.io
deletionPolicy: Delete
```

2. I created a groupsnapshot:

```
---
apiVersion: groupsnapshot.storage.k8s.io/v1beta1
kind: VolumeGroupSnapshot
metadata:
  name: new-groupsnapshot-demo
spec:
  source:
    selector:
      matchLabels:
        # The PVCs will need to have this label for it to be
        # included in the VolumeGroupSnapshot
        app.kubernetes.io/name: postgresql
  volumeGroupSnapshotClassName: csi-hostpath-groupsnapclass
```

3. I noticed that although groupsnapshot is `readyForUse`, `volumeSnapshot` object has errors:

```
spec:
  source:
    persistentVolumeClaimName: csi-pvc
status:
  boundVolumeSnapshotContentName: snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f
  error:
    message: Failed to set default snapshot class with error cannot find default snapshot
      class
    time: "2024-11-20T19:41:56Z"
  readyToUse: false
```

And there were errors in snapshot-controller logs:

```
    snapshot-controller : I1120 14:38:35.463359  867688 util.go:284] storeObjectUpdate updating snapshot "default/snapshot-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f" with version 8262                                                   csi-snapshotter : I1120 14:38:35.517898  867697 util.go:284] storeObjectUpdate updating content "snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f" with version 816
csi-snapshotter : I1120 14:38:35.517938  867697 snapshot_controller.go:321] createSnapshotWrapper: Creating snapshot for content snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f through the plugin ...
csi-snapshotter : E1120 14:38:35.517944  867697 snapshot_controller.go:239] failed to getCSISnapshotInput snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f without a snapshot class                                              csi-snapshotter : I1120 14:38:35.517957  867697 snapshot_controller.go:169] updateContentStatusWithEvent[snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f]: the same error &{2024-11-20 14:15:11 -0500 EST 0xc000585280} is alrea
dy set                                                                                                                                                                                                                                                       csi-snapshotter : E1120 14:38:35.517968  867697 snapshot_controller.go:147] checkandUpdateContentStatus [snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f]: error occurred failed to get input parameters to create snapshot for
content snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f: "failed to take snapshot snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f without a snapshot class"
csi-snapshotter : E1120 14:38:35.517978  867697 snapshot_controller_base.go:357] could not sync content "snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f": failed to get input parameters to create snapshot for content snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f: "failed to take snapshot snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f without a snapshot class"
csi-snapshotter : I1120 14:38:35.517985  867697 snapshot_controller_base.go:228] Failed to sync content "snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f", will retry again: failed to get input parameters to create snapshot f
or content snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f: "failed to take snapshot snapcontent-cd7955d4ac3c780f7982722d67de67a737fb5c6242b64d460a8c73a14903357f without a snapshot class"

```

### PR DESCRIPTION
**Pull request by leonardoce**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1203

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by leonardoce** [comment]
Please just close it if I don't understand the context.
In my opinion, we can reach createSnapshotWrapper for group snapshot members when we don't have a snapshot class.
In that case, we should not panic.

File: pkg/sidecar-controller/snapshot_controller.go [hunk 1]
```diff
@@ -339,7 +339,12 @@ func (ctrl *csiSnapshotSideCarController) createSnapshotWrapper(content *crdv1.V
 		return content, fmt.Errorf("failed to add VolumeSnapshotBeingCreated annotation on the content %s: %q", content.Name, err)
 	}
 
-	parameters, err := utils.RemovePrefixedParameters(class.Parameters)
+	var parameters map[string]string
```
Comment (xing-yang) [inline pkg/sidecar-controller/snapshot_controller.go:342]: I think we should not call createSnapshotWrapper() if it is a member of a dynamically provisioned volume groupsnapshot.  I need to look into more and see where is the best place to do that check.

Comment (leonardoce) [reply pkg/sidecar-controller/snapshot_controller.go:342]: I'm proposing another fix. I'm trying another fix; let's see if it is more reasonable.

**Comment by leonardoce** [comment]
The only other call to `createSnapshotWrapper` is in `createSnapshot`, and we already don't call it for group snapshot members: 

https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/sidecar-controller/snapshot_controller.go#L89-L93

**Comment by gnufied** [comment]
So while this indeed does make the panic disappear. It does not actually fixes underlying issue. 

```
  boundVolumeSnapshotContentName: snapcontent-ea4accb874afdb52cfe72416e09a4ec35293f5c191499af491d3a1ca3982d42d
  error:
    message: Failed to set default snapshot class with error cannot find default snapshot
      class
    time: "2024-11-22T21:25:52Z"
```

What is worse is, when this happens - snapshotter is spamming the API server with `volumesnapshot` updates. See - 

```
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  creationTimestamp: "2024-11-22T21:04:18Z"
  finalizers:
  - snapshot.storage.kubernetes.io/volumesnapshot-in-group-protection
  - snapshot.storage.kubernetes.io/volumesnapshot-as-source-protection
  generation: 1
  labels:
    groupsnapshot.storage.k8s.io/volumeGroupSnapshotName: new-groupsnapshot-demo
  name: snapshot-ea4accb874afdb52cfe72416e09a4ec35293f5c191499af491d3a1ca3982d42d
  namespace: default
  resourceVersion: "7599"
  uid: 721e6795-4c7d-4516-b6d4-79da6ed0cf01
spec:
  source:
    persistentVolumeClaimName: csi-pvc
status:
  boundVolumeSnapshotContentName: snapcontent-ea4accb874afdb52cfe72416e09a4ec35293f5c191499af491d3a1ca3982d42d
  readyToUse: false
---
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  creationTimestamp: "2024-11-22T21:04:18Z"
  finalizers:
  - snapshot.storage.kubernetes.io/volumesnapshot-in-group-protection
  - snapshot.storage.kubernetes.io/volumesnapshot-as-source-protection
  generation: 1
  labels:
    groupsnapshot.storage.k8s.io/volumeGroupSnapshotName: new-groupsnapshot-demo
  name: snapshot-ea4accb874afdb52cfe72416e09a4ec35293f5c191499af491d3a1ca3982d42d
  namespace: default
  resourceVersion: "7600"
  uid: 721e6795-4c7d-4516-b6d4-79da6ed0cf01
spec:
  source:
    persistentVolumeClaimName: csi-pvc
status:
  boundVolumeSnapshotContentName: snapcontent-ea4accb874afdb52cfe72416e09a4ec35293f5c191499af491d3a1ca3982d42d
  error:
    message: Failed to set default snapshot class with error cannot find default snapshot
      class
    time: "2024-11-22T21:26:30Z"
```

This objects's `resourceVersion` is 7600. That means, this object has been updated 7600 times in just last 10 minutes or so. This seems pretty weird.

**Comment by leonardoce** [comment]
I agree with you, @gnufied, but unfortunately, I cannot reproduce the behavior you're experiencing.
I can provision it successfully without any volumesnapshotclass.

**Comment by gnufied** [comment]
You are right that `resourceVersion` doesn't start from 1. But I see very frequent updates to `volumesnapshot` object:

```
snapshot-controller : I1122 16:55:09.550074 1559484 util.go:284] storeObjectUpdate updating snapshot "default/snapshot-e10e58934be6b25e706ea1cc371c18f08e4638bcff4c001b8f294fbaf39f9241" with version 9032
snapshot-controller : I1122 16:55:09.747271 1559484 request.go:632] Waited for 197.153734ms due to client-side throttling, not priority and fairness, request: GET:https://192.168.1.108:6443/apis/snapshot.storage.k8s.io/v1/namespaces/default/volumesnapsh
ots/snapshot-e10e58934be6b25e706ea1cc371c18f08e4638bcff4c001b8f294fbaf39f9241
snapshot-controller : I1122 16:55:09.946757 1559484 request.go:632] Waited for 198.276968ms due to client-side throttling, not priority and fairness, request: PUT:https://192.168.1.108:6443/apis/snapshot.storage.k8s.io/v1/namespaces/default/volumesnapshots/snapshot-e10e58934be6b25e706ea1cc371c18f08e4638bcff4c001b8f294fbaf39f9241/status
snapshot-controller : I1122 16:55:09.949437 1559484 util.go:284] storeObjectUpdate updating snapshot "default/snapshot-e10e58934be6b25e706ea1cc371c18f08e4638bcff4c001b8f294fbaf39f9241" with version 9033                                                   snapshot-controller : I1122 16:55:10.147201 1559484 request.go:632] Waited for 197.713516ms due to client-side throttling, not priority and fairness, request: GET:https://192.168.1.108:6443/apis/snapshot.storage.k8s.io/v1/namespaces/default/volumesnapsh
ots/snapshot-e10e58934be6b25e706ea1cc371c18f08e4638bcff4c001b8f294fbaf39f9241                                                                                                                                                                                snapshot-controller : I1122 16:55:10.347235 1559484 request.go:632] Waited for 198.449458ms due to client-side throttling, not priority and fairness, request: PUT:https://192.168.1.108:6443/apis/snapshot.storage.k8s.io/v1/namespaces/default/volumesnapsh
ots/snapshot-e10e58934be6b25e706ea1cc371c18f08e4638bcff4c001b8f294fbaf39f9241/status
snapshot-controller : I1122 16:55:10.350431 1559484 util.go:284] storeObjectUpdate updating snapshot "default/snapshot-e10e58934be6b25e706ea1cc371c18f08e4638bcff4c001b8f294fbaf39f9241" with version 9036                                                   snapshot-controller : I1122 16:55:10.381989 1559484 util.go:284] storeObjectUpdate updating groupsnapshot "default/new-groupsnapshot-demo" with version 9038
```

Which is concerning for me.

**Comment by gnufied** [comment]
okay, this is looking better now:

```
apiVersion: v1
items:
- apiVersion: snapshot.storage.k8s.io/v1
  kind: VolumeSnapshot
  metadata:
    creationTimestamp: "2024-11-22T22:14:35Z"
    finalizers:
    - snapshot.storage.kubernetes.io/volumesnapshot-in-group-protection
    - snapshot.storage.kubernetes.io/volumesnapshot-as-source-protection
    generation: 1
    labels:
      groupsnapshot.storage.k8s.io/volumeGroupSnapshotName: new-groupsnapshot-demo
    name: snapshot-bc3e5c8c83477504cb366a854e59b506ffa996fe652cfb58f9a93cd520640fb5
    namespace: default
    resourceVersion: "428"
    uid: 846e12d1-3d37-4721-9cee-205169df0959
  spec:
    source:
      persistentVolumeClaimName: csi-pvc
  status:
    boundVolumeSnapshotContentName: snapcontent-bc3e5c8c83477504cb366a854e59b506ffa996fe652cfb58f9a93cd520640fb5
    creationTime: "2024-11-22T22:14:35Z"
    readyToUse: true
    restoreSize: 1Gi
    volumeGroupSnapshotName: new-groupsnapshot-demo
kind: List
metadata:
  resourceVersion: ""
```

File: pkg/sidecar-controller/snapshot_controller.go [hunk 2]
```diff
@@ -315,7 +315,13 @@ func (ctrl *csiSnapshotSideCarController) checkandUpdateContentStatusOperation(c
...
 		return updatedContent, nil
 	}
-	return ctrl.createSnapshotWrapper(content)
+
+	_, groupSnapshotMember := content.Labels[utils.VolumeGroupSnapshotHandleLabel]
+	if !groupSnapshotMember {
+		return ctrl.createSnapshotWrapper(content)
+	}
+
+	return content, nil
```
Comment (gnufied) [inline pkg/sidecar-controller/snapshot_controller.go:324]: Should we have some tests for this? For now - unit tests will be good. In long term, we should have some e2e as well I think. I am not sure if all users will have a default snapshotclass when testing volumegroupsnapshot feature.

Comment (xing-yang) [reply pkg/sidecar-controller/snapshot_controller.go:324]: Opened an issue to track this: https://github.com/kubernetes-csi/external-snapshotter/issues/1211

**Pull request merged**